services/wallet/balance/balance_cache.go

```diff
@@ -35,145 +36,176 @@ type Cacher interface {
 
 // Interface for cache of balances.
 type CacheIface interface {
-	GetBalance(account common.Address, blockNumber *big.Int) *big.Int
-	GetNonce(account common.Address, blockNumber *big.Int) *int64
-	AddBalance(account common.Address, blockNumber *big.Int, balance *big.Int)
-	AddNonce(account common.Address, blockNumber *big.Int, nonce *int64)
+	GetBalance(account common.Address, chainID uint64, blockNumber *big.Int) *big.Int
+	GetNonce(account common.Address, chainID uint64, blockNumber *big.Int) *int64
+	AddBalance(account common.Address, chainID uint64, blockNumber *big.Int, balance *big.Int)
+	AddNonce(account common.Address, chainID uint64, blockNumber *big.Int, nonce *int64)
 	Clear()
 }
 
+type balanceCacheType map[common.Address]map[uint64]map[uint64]*big.Int      // address->chainID->blockNumber->balance
+type nonceCacheType map[common.Address]map[uint64]map[uint64]*int64          // address->chainID->blockNumber->nonce
+type nonceRangesCacheType map[common.Address]map[uint64]map[int64]nonceRange // address->chainID->blockNumber->nonceRange
+type sortedNonceRangesCacheType map[common.Address]map[uint64][]nonceRange   // address->chainID->[]nonceRange
+
 type Cache struct {
 	// balances maps an address to a map of a block number and the balance of this particular address
-	balances     map[common.Address]map[uint64]*big.Int // we don't care about block number overflow as we use cache only for comparing balances when fetching, not for UI
-	nonces       map[common.Address]map[uint64]*int64   // we don't care about block number overflow as we use cache only for comparing balances when fetching, not for UI
-	nonceRanges  map[common.Address]map[int64]nonceRange
-	sortedRanges map[common.Address][]nonceRange
+	balances     balanceCacheType
+	nonces       nonceCacheType
+	nonceRanges  nonceRangesCacheType
+	sortedRanges sortedNonceRangesCacheType
 	rw           sync.RWMutex
 }
 
 func NewCache() *Cache {
 	return &Cache{
-		balances:     make(map[common.Address]map[uint64]*big.Int),
-		nonces:       make(map[common.Address]map[uint64]*int64),
-		nonceRanges:  make(map[common.Address]map[int64]nonceRange),
-		sortedRanges: make(map[common.Address][]nonceRange),
+		balances:     make(balanceCacheType),
+		nonces:       make(nonceCacheType),
+		nonceRanges:  make(nonceRangesCacheType),
+		sortedRanges: make(sortedNonceRangesCacheType),
 	}
 }
 
 func (b *Cache) Clear() {
 	b.rw.Lock()
 	defer b.rw.Unlock()
 
-	for address, cache := range b.balances {
-		if len(cache) == 0 {
+	for address, chainCache := range b.balances {
+		if len(chainCache) == 0 {
 			continue
 		}
 
-		var maxBlock uint64 = 0
-		var minBlock uint64 = 18446744073709551615
-		for key := range cache {
-			if key > maxBlock {
-				maxBlock = key
+		for chainID, cache := range chainCache {
+			if len(cache) == 0 {
+				continue
 			}
-			if key < minBlock {
-				minBlock = key
+
+			var maxBlock uint64 = 0
+			var minBlock uint64 = 18446744073709551615
```

Comment:
I wonder where is this constant coming from. Can it be sourced for example, from MaxUint64 from a reliable source?